Gems/Atom/RPI/Code/Source/RPI.Public/Shader/Shader.cpp

```diff
@@ -188,15 +188,22 @@ namespace AZ
             }
 
             ShaderVariantFinderNotificationBus::Handler::BusConnect(m_asset.GetId());
-            Data::AssetBus::Handler::BusConnect(m_asset.GetId());
 
+            m_reloadedAssets.clear();
+            const auto& supervariants = m_asset->GetCurrentShaderApiData().m_supervariants;
+            m_expectedAssetReloadCount = 1 /*m_asset*/ + supervariants.size();
```

Comment:
Is this the only place this variable (m_expectedAssetReloadCount) is updated? Shouldn't it be updated within OnAssetReloaded as well?

Reply:
This is the only place where it should be updated. The key is to detect the moment when `m_reloadedAssets.size() == m_expectedAssetReloadCount` (Which is tested during `OnAssetReloaded()`.